### ISSUE TITLE
Problems converting Java.ArrayList<Float> to NDArray in scala.

### ISSUE THREAD

**Issue by borissmidt** [deeplearning4j/nd4j#1991]
#### Issue Description
I'm trying to convert a Java.util.List<java.Float> into a NDArray:
I have the following scala code:
```
val vector: util.List[Float] = model.getVector("word")
val scalaArray = vector.asScala.toArray
```
Then if i try to convert it, with `scalaArray.toNDArray()` i with the implicit function it doesn't find the implicit conversion:
Then i tried then i tells me that it cannot convert java.Float to an NDArray, since this is a java object But scala has no real way to convert it explicitly to an:
```
floatArray2INDArray(scalaArray).toNDArray()
```
and even this:
```
floatArray2INDArray(scalaArray.map(_.toFloat)).toNDArray()
```
Eventually i just solved it in java code like this:
```
    public INDArray getWordVector(String word){
        float[] buffer = new float[model.getDim()];
        List<Float> vector = model.getVector(word);
        for(int i = 0;i < model.getDim() ;i++){
            buffer[i] = vector.get(i);
        }
        return Nd4j.create(buffer);
    }
```
a scala oneliner solution is where to option allows you to unbox the Float object into a primitive float:
```
    val scalaArray = vector.asScala.toArray.flatMap(float => Option(float).map(_.toFloat)).toNDArray
```
- expected behavior
Direct conversion functions between Seq/ List containers To NDArrays. Since most java/ scala code operate on List/Seq and not the Array.

- encountered behavior

It is hard to convert from a boxed java.Float to a scala.Float in scala. 
There are no functions in java to from a java.List to Float[] without manually copying.

I understand that this is a problem on Scala its side because opaque boxing and unboxing of a Float.
On the java side it is the missing functions to convert between Java List to Array and functions to go from java Float[] to float[]

#### Version Information

Please indicate relevant versions, including, if relevant:

* Deeplearning4j version
ND4j 0.8.0
ND4s 0.8.0
#### Contributing

If you'd like to help us fix the issue by contributing some code, but would
like guidance or help in doing so, please mention it!
Since i could write convenience code to do the conversions from the scala types and the java list. But these probably have to be documented that they are rather inefficient because of the way they work.

**Comment by saudet** [comment]
That's a general limitation in Java as well. Since we need to copy the values from arrays anyway, the equivalent with List wouldn't be that inefficient. Additional factory methods are welcome! Please send your pull request. :)

**Comment by borissmidt** [comment]
I will be right on it and try to implement it into a rather generic way. :) So if i see it correctly the right way will be to copy the values from the list directly into an NDArray without creating an intermedian array.

**Comment by agibsonccc** [comment]
@borissmidt you can have that just be zero if you want. I'm more inclined to treat this similarly to a float array in java.

**Comment by borissmidt** [comment]
@agibsonccc I see 2 problems for which i will have to find a solution, 
The first one is Type erasure, so i'm unable to overload the function for a List<Float> and List<Double> since the jvm will be unable to see the difference after compilation. 

Another problem is that these functions will have to be implemented in the abstract factory but also in the BaseNdArrayFactory and its childeren.

To solve the second problem, i could just copy the List to an array and then delegate the call to the usual functions.

**Comment by borissmidt** [comment]
I already created a pull request that solves this problem for scala. But i still have to tackle the java case.

**Comment by borissmidt** [comment]
@agibsonccc, I'm still interested to implement these conversion functions, But i have to give them different names to avoid type erasure of the java.util.List<X>.  Is this a reasonable solution or do you have a better idea to do this?

My proposed solution would be to just copy it to a primitive array and then use the existing functions. Since this would only cause in a small change.

**Comment by agibsonccc** [comment]
Yeah that would be fine.

**Comment by huitseeker** [comment]
In Scala, you can just:

```

scala> import scala.collection.breakOut
import scala.collection.breakOut
scala> import scala.collection.JavaConverters._
import scala.collection.JavaConverters._
scala> def f(l: java.util.List[java.lang.Float]): Array[java.lang.Float] = l.asScala.map(identity) (breakOut)
f: (l: java.util.List[Float])Array[Float]
```

Unboxing should not be needed if you avoid boxing, but if you don't there is an unsafe `Float.box` you can use in cases you are certain you boxed.

**Comment by saudet** [comment]
That's not the issue, AFAIU. We need a float[], but we're getting a Float[] or something. Unboxing all elements isn't very efficient.

**Comment by borissmidt** [comment]
The original problem is just that most of the java works use the default collections like List. But java its collections can only contain elements of type object, But the primitives are not objects. Thus you cannot go from List<Float> to a primitive array. While List<Float> is used by many java libraries but float[] isn't.
Thus it would be useful for the library to have convenience functions to convert from List<Float> to NDFloatArray.

**Comment by huitseeker** [comment]
So, rather than speaking besides the point I'd rather equip you with tools to make things happen.

In Scala, Java Arrays are Scala ones too:

```
scala> import java.util.Arrays
import java.util.Arrays
scala> var m = scala.Array(8,7,6,5,4,3) 
m: Array[Int] = Array(8, 7, 6, 5, 4, 3)
scala> Arrays.sort(m)
scala> m
res2: Array[Int] = Array(3, 4, 5, 6, 7, 8)
```

in Scala, `scala.Float`, a primitive value type, is convertible to a `java.lang.Float` through an implicit conversion in `Predef` (everybody gets these implicits):
https://github.com/scala/scala/blob/2.12.x/src/library/scala/Predef.scala#L455

Boxed `java.lang.Float` objects are auto-convertible to a `scala.Float`:
https://github.com/scala/scala/blob/2.12.x/src/library/scala/Predef.scala#L472

And `breakout` lets you implicitly invoke the basic collection construction for translating without adding one more traversal to your collection construction (as long as there's one). 

So, *in scala*, as this issue started out with, you get:
```
scala> def f(l: List[java.lang.Float]): Array[Float] = l.map((x)=>x:Float) (breakOut)
f: (l: List[Float])Array[Float]
```

(which implicitly applies `Float2float`)

And coming from a java list:
```
scala> import scala.collection.JavaConverters
import scala.collection.JavaConverters

scala> def f(l: java.util.List[java.lang.Float]): Array[Float] = l.asScala.map((x)=>x:Float) (breakOut)
f: (l: java.util.List[Float])Array[Float]
```

(the map is here because it's still applying `Float2float`) 

If you want to do this in Java, you'll need to unbox the `java.lang.Float` elements while filling your `float[]`, because, well, that's all you have. I wasn't saying anything more controversial that : unbox if all you have is boxed objects, and take the primitive elements otherwise.

**Comment by borissmidt** [comment]
So should i add the convenience functions to the library? But  i'm still confused on how to handle the `null` pointer. 
Is it a yes i would like to have a null replaced by 0. Or is an exception tolerable since the scala implementation would just throw the exception?

**Comment by huitseeker** [comment]
You should add those convenience functions. My opinion is you should throw on null, both in Java and Scala. The functions I have above will throw on null, but will have tremendously better performance than ones which null-check with `Option` (the object creation alone is intractable at the data volumes we're envisioning).

In Scala, your implementation can look like the above, in Java, it should look like the similar convenience function from Guava https://github.com/google/guava/blob/master/guava/src/com/google/common/primitives/Floats.java#L414
(in fact, if you look around, you'll see we have Guava in dependencies, and you can use just that).

**Comment by TamouzeAssi** [comment]
Sir @raver119 
I tired in Jython to create a double array:
`myArray = jarray.array(vector, java.lang.Double), 
where vector is  <type 'java.util.ArrayList'>. the type of myArray is <type 'array.array'>. So when i pass it to the create function like that `nd = Nd4j.create(myArray)`, i get the following exception:
 nd = Nd4j.create(myArray)
TypeError: create(): 1st arg can't be coerced to int[], float[], double[][], float[][], org.nd4j.linalg.api.buffer.DataBuffer, int, double[]

**Comment by huitseeker** [comment]
Fixed with merge of #2043
